src/shadow-embed.js

```diff
@@ -47,3 +56,85 @@ export function createShadowEmbedRoot(hostElement, extensionIds) {
 
   return shadowRoot;
 }
+
+
+/**
+ * Transforms CSS to isolate AMP CSS within the shadow root and reduce the
+ * possibility of high-level conflicts. There are two types of transformations:
+ * 1. Root transformation: `body` -> `amp-body`, etc.
+ * 2. Scoping: `a {}` -> `#i-amp-sd-123 a {}`.
+ *
+ * @param {!ShadowRoot} shadowRoot
+ * @param {string} css
+ * @return {string}
+ * @visibleForTesting
+ */
+export function scopeShadowCss(shadowRoot, css) {
+  const id = dev().assert(shadowRoot.id);
+  const doc = shadowRoot.ownerDocument;
+  let rules = null;
+  // Try to use a separate document.
+  try {
+    rules = getStylesheetRules(doc.implementation.createHTMLDocument(''), css);
+  } catch (e) {
+    // Ignore.
+  }
+  // Try to use the current document.
+  if (!rules) {
+    try {
+      rules = getStylesheetRules(doc, css);
+    } catch (e) {
+      // Ignore.
+    }
+  }
+
+  // No rules could be parsed - return css as is.
+  if (!rules) {
+    return css;
+  }
+
+  // Patch selectors.
+  return ShadowCSS.scopeRules(rules, `#${id}`, transformRootSelectors);
+}
+
+
+/**
+ * Replaces top-level selectors such as `html` and `body` with their polyfill
+ * counterparts: `amp-html` and `amp-body`.
+ * @param {string} selector
+ * @return {string}
+ */
+function transformRootSelectors(selector) {
+  return selector.replace(/(html|body)/g, function(match, name, pos) {
```

Comment:
@jridgewell Need to merge this now. But could you please advise on this and I'll send a separate PR. I hope my comments on regex clarify other questions.